### ISSUE TITLE
Multiple genes

### ISSUE THREAD

**Issue by Gregory94** [gacatag/IntEREst#12]
Hello,

I want to run IntEREst  (v1.14.0) for more than one gene. Currently I'm using the following code for that (largely taken from [the IntEREst tutorial](https://bioconductor.org/packages/release/bioc/vignettes/IntEREst/inst/doc/IntEREst.html)):

```
suppressMessages(library("IntEREst"))

###OPTIONS###
gene <- "all" #EITHER USE GENE NAME OR SET TO 'all' WHEN ALL GENES NEED TO BE ANALYSED
outDir<- file.path("path/out") #WHERE THE OUTPUT FILES SHOULD BE STORED
bamF <- "path/test.bam"
mutType <- "ZRSR2mut"
Control <- "test"
processNReads <- 10000
cores <- 1
#############

sampleNames <- tools::file_path_sans_ext(basename(bamF))#"small_test_SRR1691637_ZRSR2Mut_RHBDD3" 

if(strsplit(basename(bamF), split="\\.")[[1]][-1]=="rds"){
  design_table <- readRDS(bamF)
  bamFiles <- design_table$filename
  status <- design_table$ZRSR2_status
}else if(strsplit(basename(bamF), split="\\.")[[1]][-1]=="bam"){
  bamFiles <- c(bamF)
  status <- "None"
}else{
  stop("No valid file selected. Enter path to either .rds file including list of bam files or .bam file")
}

# IMPORTING GENOME
BSgenome.Hsapiens.UCSC.hg19 <-
  BSgenome.Hsapiens.UCSC.hg19::BSgenome.Hsapiens.UCSC.hg19

if(tolower(gene)=='all'){
  geneName <- unique(u12$gene_name)
}else{
  geneName <- gene
}

###CREATING A REFERENCE DATASET (in GFF3 FILE FORMAT) WITH ALL EXON LOCATIONS OF THE GENES OF INTEREST
print('Creating reference dataset ...')

# Getting information (i.e. intron and exon locations) related to gene of interest
ref<-u12[u12[,"gene_name"] %in% geneName,]
# Extracting exons only
tmpEx<-ref[ref[,"int_ex"]=="exon",]
# Building GFF3 file
exonDat<- cbind(tmpEx[,3], ".",
                tmpEx[,c(7,4,5)], ".", tmpEx[,6], ".",paste("ID=exon",
                                                            tmpEx[,11], "; Parent=ENST00000413811", sep="") )
trDat<- c(tmpEx[1,3], ".", "mRNA", as.numeric(min(tmpEx[,4])),
          as.numeric(max(tmpEx[,5])), ".", tmpEx[1,6], ".",
          "ID=ENST00000413811")
gff3File<-paste(outDir, "gffFile.gff", sep="/")
cat("##gff-version 3\n",file=gff3File, append=FALSE)
cat(paste(paste(trDat, collapse="\t"),"\n", sep=""),
    file=gff3File, append=TRUE)
write.table(exonDat, gff3File,
            row.names=FALSE, col.names=FALSE,
            sep='\t', quote=FALSE, append=TRUE)
# Extracting U12 introns info from 'u12' data
u12Int<-u12[u12$int_ex=="intron"&u12$int_type=="U12",]

# Building reference which includes all intron and exon positions for the selected gene.
testRef<- referencePrepare(sourceBuild="file",
                           filePath=gff3File, u12IntronsChr=u12Int[,"chr"],
                           u12IntronsBeg=u12Int[,"begin"],
                           u12IntronsEnd=u12Int[,"end"], collapseExons=TRUE,
                           fileFormat="gff3", annotateGeneIds=FALSE)

```
This works fine as long as I take a single gene (e.g. `gene <- "RHBDD3"` yields the same result as when I literally run the code from the tutorial). But when I want to use all genes featuring u12 introns (e.g. using the command `unique(u12$gene_name)`), I get the following error when using the `referencePrepare` function:

```
Error in matrix(as.vector(unlist(sapply(exons, function(tmp) return(c((as.character(GenomicRanges::seqnames(tmp)))[1],  : 
  'data' must be of a vector type, was 'NULL'
In addition: Warning message:
In makeTxDbFromGRanges(gr, metadata = metadata) :
  The following transcripts were dropped because their exon ranks could not be inferred (either because the exons are not on the same chromosome/strand or because they
  are not separated by introns): ENST00000413811
```

I suspect it is something obvious I am overlooking, but I'm not quite sure how to solve this. Do you have any suggestions?

Thanks in advance!

Best,
Gregory

**Comment by gacatag** [comment]
Hi @Gregory94  !

Without seeing your codes, it would be a bit difficult for me to know exactly why you are getting this error. However, s.th. that may be causing the error is the format of the GFF file that you are creating ! In the example (with one gene only) I had defined the transcript (i.e "ENST00000413811") and its exons such that they are valid for the gene RHBDD3. If you are including other genes you would need to define their transcript IDs and exons correctly for the genes. You can see the contents of the GFF file using this command in R:
 
`file.show(gff3File) ` 

I hope this is helpful !

Best regards,

Ali

**Comment by Gregory94** [comment]
Hi Ali,

Thank you for your reply! The code I provided in my question is all that I am using until the error arises. I can disable the `referencePrepare` function and then the codes works (including all the downstream functions like the U12 annotation and the `interest` function), but I am not entirely certain what influence disabling this function has on the results?
Do I understand correctly that when I create the gff-file, I only need to update the parent transcript to match that of the gene of interest (e.g. the transcript ID's that are present in the `u12` data)? If yes, may I then ask why you're using ENST00000413811 as transcript for RHBDD3, because this isn't one of its transcripts, right (e.g. see also [ensembl](https://www.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000100263;r=22:29259872-29268209))?

Best,
Gregory

**Comment by gacatag** [comment]
Hi again @Gregory94 !
Sorry for the delay in response ! Yes, you are right. It seems that the name of the gene doesn't match the transcript and needs to be corrected. Thank you for pointing that out. At some point, I changed the transcript that I used in the example but probably forgot to update the gene name.

Regarding the reference, however, please note that in the vignette I wanted to have an example of using the `referencePrepare` function which executes fast, therefore I generated a very small test GFF file to build a reference based on. In practice, I assume that a user would either want to use gene annotations from other `sourceBuild` such as "UCSC" or "biomaRt" or that she/he would use an already available gene annotation (i.e. GFF file).

Cheers,

Ali

**Comment by Gregory94** [comment]
Hi Ali,

Thank you for the clarification! I am now using the 'UCSC' sourceBuild in `referencePrepare` and that works!
However, this only seems to work for hg19, whilst I prefer to use hg38 if possible. To do this I use the following `referencePrepare` function:
```
referencePrepare (sourceBuild="UCSC", ucscGenome="hg38",
     addCollapsedTranscripts=TRUE, ignore.strand=FALSE, ucscTableName="refGene")
```
and for the `annotateU12` function I'm using:
```
annotateU12(pwmU12U2=list(pwmU12db[[1]][,11:17],pwmU12db[[2]],
                            pwmU12db[[3]][,38:40],pwmU12db[[4]][,11:17],
                            pwmU12db[[5]][,38:40]),
              pwmSsIndex=list(indexDonU12=1, indexBpU12=1, indexAccU12=3,
                              indexDonU2=1, indexAccU2=3),
              referenceChr=u12[ind,'chr'],
              referenceBegin=u12[ind,'begin'],
              referenceEnd=u12[ind,'end'],
              referenceIntronExon=u12[ind,"int_ex"],
              intronExon="intron",
              matchWindowRelativeUpstreamPos=c(NA,-29,NA,NA,NA),
              matchWindowRelativeDownstreamPos=c(NA,-9,NA,NA,NA),
              minMatchScore=c(rep(paste(80,"%",sep=""),2), "40%",
                              paste(80,"%",sep=""), "40%"),
              refGenome=BSgenome.Hsapiens.UCSC.hg38,
              setNaAs="U2",
              annotateU12Subtype=TRUE)
```
where BSgenome.Hsapiens.UCSC.hg38 is loaded from `BSgenome.Hsapiens.UCSC.hg38::BSgenome.Hsapiens.UCSC.hg38`.

This gives me the following warnings and error:

```
Warning messages:
1: In .extract_cds_locs_from_UCSC_txtable(ucsc_txtable) :
  UCSC data anomaly in 880 transcript(s): the cds cumulative length is
  not a multiple of 3 for transcripts ‘NM_012314’ ‘NM_006781’
  ‘NM_001281972’ ‘NM_001102594’ ‘NM_001256512’ ‘NM_001289154’
  ‘NM_001286475’ ‘NM_182647’ ‘NM_001202523’ ‘NM_001170738’
  ‘NM_001322841’ ‘NM_006781’ ‘NM_181721’ ‘NM_002457’ ‘NM_001287810’
  ‘NM_001102596’ ‘NM_001355465’ ‘NM_001282170’ ‘NM_001322371’
  ‘NM_001322371’ ‘NM_001322371’ ‘NM_001170794’ ‘NM_004829’
  ‘NM_001098721’ ‘NM_001794’ ‘NM_007028’ ‘NM_004522’ ‘NM_001322841’
  ‘NM_004557’ ‘NM_001145458’ ‘NM_001145458’ ‘NM_001005742’
  ‘NM_001200019’ ‘NM_001329128’ ‘NM_001318399’ ‘NM_014218’ ‘NM_020469’
  ‘NM_001243965’ ‘NM_006806’ ‘NM_001103184’ ‘NM_001103184’
  ‘NM_001103184’ ‘NM_001184961’ ‘NM_001008747’ ‘NM_001348286’
  ‘NM_001083539’ ‘NM_005155’ ‘NM_015296’ ‘NM_0 [... truncated]
2: In makeTxDb(transcripts, splicings, genes = genes, chrominfo = chrominfo,  :
  3 transcripts were dropped because they are on unknown sequences
  (transcripts NM_001354832, NM_001384980, NM_001384981)
3: In .set_group_names(grl, use.names, txdb, by) :
  some group names are NAs or duplicated
[1] "Annotating u12 introns ..."
Error in .Call2("C_solve_user_SEW", refwidths, start, end, width, translate.negative.coord,  :
  solving row 1288: 'allow.nonnarrowing' is FALSE and the supplied start (51207303) is > refwidth + 1
Calls: annotateU12
Execution halted
```
Are there other things I need to do before I can use the hg38 genome? I have seen [another issue](https://github.com/gacatag/IntEREst/issues/3) about a similar topic, but I am not entirely sure how this relates to this situation.

Best,
Gregory

**Comment by gacatag** [comment]
Hi @Gregory94 !

I'm  not getting any error when I run your codes! However, `u12` is a test data based on HG19 ! therefore no U12 type introns are detected ! 
you need to input the columns from the reference that you built via `referencePrepare()` command to the `referenceChr`, `referenceBegin`, `referenceEnd`, `referenceIntronExon` parameters of the `annotateU12()` function.

Please see the manual for `referencePrepare` .

Cheers,

**Comment by Gregory94** [comment]
Hi Ali,

Thanks again for your help! I changed the parameters from the `annotateU12()` function to read the output of the `referencePrepare()` instead of the `U12` testdata and that worked! It's obvious in hencesight, but it took forever for the `annotateU12` function to finish when considering the whole genome and before I thought the code was stuck somewhere and therefore I aborted it (apparently I need to have some more patience).

Best,
Gregory
